src/home/homectl.c

```diff
@@ -4054,6 +4053,197 @@ static int redirect_bus_mgr(void) {
+                                               "Failed to activate home dir, even after %u tries.", n_tries);
+
+                /* Let's start by checking if this all is even necessary, i.e. if the useFallback boolean field is actually set. */
+                r = bus_call_method(bus, bus_mgr, "GetUserRecordByName", &error, &reply, "s", NULL); /* empty user string means: our calling user */
+                if (r < 0)
+                        return log_error_errno(r, "Failed to inspect home: %s", bus_error_message(&error, r));
+
+                r = sd_bus_message_read(reply, "sbo", &json, NULL, NULL);
+                if (r < 0)
+                        return bus_log_parse_error(r);
+
+                r = json_parse(json, JSON_PARSE_SENSITIVE, &v, NULL, NULL);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse JSON identity: %m");
+
+                hr = user_record_new();
+                if (!hr)
+                        return log_oom();
+
+                r = user_record_load(hr, v, USER_RECORD_LOAD_REFUSE_SECRET|USER_RECORD_LOG|USER_RECORD_PERMISSIVE);
+                if (r < 0)
+                        return r;
+
+                if (!hr->use_fallback) /* Nice! We are done, fallback logic not necessary */
+                        break;
+
+                if (!secret) {
+                        r = acquire_passed_secrets(hr->user_name, &secret);
+                        if (r < 0)
+                                return r;
+                }
+
+                for (;;) {
+                        _cleanup_(sd_bus_message_unrefp) sd_bus_message *m = NULL;
+
+                        r = bus_message_new_method_call(bus, &m, bus_mgr, "ActivateHomeIfReferenced");
+                        if (r < 0)
+                                return bus_log_create_error(r);
+
+                        r = sd_bus_message_append(m, "s", NULL); /* empty user string means: our calling user */
+                        if (r < 0)
+                                return bus_log_create_error(r);
+
+                        r = bus_message_append_secret(m, secret);
+                        if (r < 0)
+                                return bus_log_create_error(r);
+
+                        r = sd_bus_call(bus, m, HOME_SLOW_BUS_CALL_TIMEOUT_USEC, &error, NULL);
+                        if (r < 0) {
+                                if (sd_bus_error_has_name(&error, BUS_ERROR_HOME_NOT_REFERENCED))
+                                        return log_error_errno(r, "Called without reference on home taken, can't operate.");
+
+                                r = handle_generic_user_record_error(hr->user_name, secret, &error, r, false);
+                                if (r < 0)
+                                        return r;
+
+                                sd_bus_error_free(&error);
+                        } else
+                                break;
```

Comment:
The logic here is a bit confusing.

If `ActivateHomeIfReferenced` is successful, you still loop the outer loop to check that `use_fallback` actually became false. If the method fails, we either stay in the inner loop (if we can retry auth) or completely return from the function and thus kill the SSH session.

So what's the point of the outer retry loop? It seems to only loop if `ActivateHomeIfReferenced` successfully returned but still somehow failed to do its job? When would something like this happen?

Reply:
it's again about the race: multiple clients starting/stopping sessions at the very same time, with and without reference counting. we want the user record where the fallback logic is off and with the most current data (which will only be available after successful activation). Hence we keep activating things until we get a user record that looks like it's right. we give up after 5 tries though, because if we keep looping around this it looks more like someone is fucking with us then actual concurrent operations...

i mean, as long as all logins and other attempts to access home dirs go through this pam-systemd-home code paths then it should in theory suffice to just do get-record→activate→get-record and be done with it. But it seemed safer to me to try a couple of times, since there's als ActivateHome without refcounting after all...